### PR TITLE
Fix calculation of smeared MET on MiniAOD

### PR DESCRIPTION
**Pull request by michaelwassmer**
#### PR description:

This pull request fixes the incorrect calculation of the nominal smeared missing transverse energy (MET). The underyling problem is the application of the JER correction factors. These factors are applied in PhysicsTools/PatUtils/interface/SmearedJetProducerT.h but later not considered when reverting to the uncorrected jets in  JetMETCorrections/Type1MET/interface/PFJetMETcorrInputProducerT.h.
This results in the problem that the JER correction factors in the smeared MET calculation effectively cancels and therefore the smeared MET is basically the same as the regular type1-corrected MET. After the changes shown below, the nominal smeared MET distributions from MiniAOD are similar to the ones obtained from NanoAOD.

This pull request includes the following:

PhysicsTools/PatUtils/plugins/SmearedJetProducer.cc
- adds a template function (SmearJet in SmearedJetProducer_namespace) to apply the JER
- if the type of the jet is not pat::Jet it reproduces what has been done so far (using the scaleEnergy function to apply the JER factor to the jet)
- if the jet is of type pat::Jet, it does the same but additionally adds the JER factor as a userFloat to the pat::Jet (needed to later remove JER from the jets in the corresponding MET module)

PhysicsTools/PatUtils/interface/SmearedJetProducerT.h
- use the template function explained above

PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc
- edit pat::Jet template specialization of the already existing RawJetExtractor class to consider a possible JER correction factor, which was added as a userFloat previously, during the jet correction removal  (if a JER factor was applied, it is now removed by applying 1/factor)
- pat::Jet template specialization, which retrieves a previously saved JER factor, of the template class/functor explained below

JetMETCorrections/Type1MET/interface/PFJetMETcorrInputProducerT.h
- add a template class/functor to retrieve JER correction factors
- in generic implementation it only applies 1 as a factor therefore doing nothing
- in case of pat::Jet, see above

Finally, there is also a mistake when saving the JERup and JERdown varied smearedMET. The different contributions of the nominal smeared MET correction and the JERup and JERdown variations are not separated correctly from on another. After the changes shown below, the varied smeared MET distributions from MiniAOD are similar to the ones obtained from NanoAOD.

DataFormats/PatCandidates/src/MET.cc
- to get the corrections/deltas for the JER variations of the smeared MET, the calculation now takes the up/down-varied smeared MET (px,py,...) and subtracts the contributions from the nominal smeared MET corrections (ref.dpx(),ref.dpy(),...) as well as the nominal type1-corrected MET (this->px(),this->py(),...)
- the previous implementation is a mystery to me :D

#### PR validation:

- 'runTheMatrix.py -l limited -i all' worked besides some workflows which had CMSDAS errors despite a working certificate
- 'scram b runtest' worked

### CODE REVIEW

**Comment by slava77** [comment]
@michaelwassmer 
please add a link to a presentation in a JME POG meeting with more details on the physics impact.

How old is this problem and would a backport be needed?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by michaelwassmer** [comment]
Here is a link to a presentation in the JetMET meeting comparing original and fixed calculation: https://indico.cern.ch/event/1048658/contributions/4405700/attachments/2263616/3842703/SmearedMET_Michael_Wassmer.pdf
Here another link containing plots comparing to NanoAOD (slide 11): https://indico.cern.ch/event/1077259/contributions/4530656/attachments/2312550/3935829/METreport2.pdf

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Since this touches MINI, tagging @cms-sw/xpog-l2 as well.

@michaelwassmer can you summarize from the JME perspective if the previous implementation in MINI could have affected any analyses, and if yes, to what extent?

**Comment by michaelwassmer** [comment]
Ok, I can try:
First of all, as far as I know, it was always communicated by JetMET that the smeared MET should not be used, because it was known to be incorrect. How long this was known, I don't know. Other people, involved in JetMET for a longer time than me, would need to answer this. The effect of the bug on the smeared MET distribution is effectively the cancellation of the, for smeared MET desired, JER corrections. So in the worst case, the incorrect smeared MET distribution used in an analysis would effectively be the regular type1-corrected PFMET distribution, which is afaik the recommended MET. Regarding the also incorrect JER up/down variations of the smeared MET, the effect of the bug is that the varied distributions are asymmetric/one-sided meaning that the down variation shows more or less no variation and the up variation looks as expected (however on top of the buggy nominal distribution).
In summary, I'd say a small number of even no analyses are affected due to the explicit recommendation not to use this MET.
I hope it is possible to understand what I mean, if not please let me know.

File: PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc [hunk 1]
```diff
@@ -16,17 +16,41 @@ namespace PFJetMETcorrInputProducer_namespace {
...
     reco::Candidate::LorentzVector operator()(const pat::Jet& jet) const {
+      reco::Candidate::LorentzVector uncorrected_jet;
       if (jet.jecSetsAvailable())
-        return jet.correctedP4("Uncorrected");
+        uncorrected_jet = jet.correctedP4("Uncorrected");
       else
-        return jet.p4();
+        uncorrected_jet = jet.p4();
+      // remove JER correction factor from pat::Jets
+      if (jet.hasUserFloat("SmearFactor")) {
```
Comment (jpata) [inline PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc:32]: under what conditions in the new code will you not have a SmearFactor userfloat available? In other words, should a failed access to SmearFactor just throw an error, rather than return 1.0?

**Review by jpata — COMMENTED** [review]
Thanks for the clarifications, so we understand that it's not expected to have had a big effect on analysis.

I'm wondering if userFloat was the only reasonable way to implement this, or would it also be possible to make the smearFactor a proper class member, to benefit from compile-time checks (and future dataformat evolution)?

File: PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc (hunk 1 above)
Comment (michaelwassmer) [reply PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc:32]: In general I agree that there is probably a more elegant solution of achieving this and I thought about it, but in the end I couldn't really get a nice solution without changing some very elementary classes, which I was hesitant to do (also because I'm not an expert on them). Maybe I can give some background information to make this more understandable.

The module calculating the JER corrections and uncertainties for the jets is the SmearedJetProducer, which is a template class defined for:
- reco::CaloJet
- reco::PFJet
- pat::Jet
This module obtains the correction factors and uncertainties and scales the jets accordingly.
So far so good. The problem now arises when trying to keep track of the different jet corrections in the MET modules relevant in this PR. As you can see, even without my changes, the PATPFJetMETcorrInputProducer producing the MET corrections already behaves differently for pat::Jets and the rest (also with template specialization). In order to obtain the uncorrected jets, different methods are necessary.
One option would probably be (though I'm not 100% sure of this) to use the methods, which are used to apply the JES corrections to the pat::Jets. These methods keep track of the different corrections. So if the JER factors would be applied the same way, the removal of the JER factor in the MET module would work automatically with the MET module as it currently is. However, the SmearedJetProducer would need to be edited to make use of these pat::Jet methods and then of course one would get the problem that these methods do not work for the reco jets, so one would have to do some specialization again.

So in summary, I totally agree that the userFloat thing is not great but I wasn't able to see the obvious solution working for all jet types. As far as I understand, one cannot avoid the specialization stuff and therefore I just used the easiest possible way exploiting the pat userFloat capability. Maybe there is an obvious solution to this, then I think we/I would need the help of an expert regarding the jet classes.

Sorry for the long reply.

Comment (jpata) [reply PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc:32]: Thanks for this summary. I think the question here is not so much elegance but rather future maintainability, as there are no guarantees about the presence or evolution of the userFloat, so essentially we can only know at runtime if things will not crash, and even then it will be difficult to tell if it's correct going forward (assuming someone else changes things in a different place).

Sorry for the naive question, but why not add a `double smearFactor` member to all the jet classes reco::CaloJet, reco::PFJet, pat::Jet that need to be smeared/unsmeared?

Then the SmearJet and RetrieveJer could be something like:

```C++
  template <>
  inline void SmearJet<pat::Jet>(pat::Jet& jet, float smearfactor) {
    jet.scaleEnergy(smearfactor);
    jet.smearFactor = smearfactor;
  }

  template <>
  class RetrieveJerT<pat::Jet> {
  public:
    RetrieveJerT() {}
    float operator()(const pat::Jet& jet) const {
      return jet.smearfactor;
  };
```

The dataformat can use a read rule to guarantee backwards compatibility (I'm not sure how this is ensured in this PR currently with the userFloat).

Comment (jpata) [reply PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc:32]: Perhaps the main question is what you should do on a missing userFloat? Return silently 1.0, or crash. In the latter case, a configuration flag like `useSmearFactor` may be necessary.

Comment (michaelwassmer) [reply PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc:32]: Hi again,
I see what you mean. Maybe the best route will be to add a member to pat::Jet, as you suggested, to store this factor and be able to compensate for it this way. This would result in the same outcome as the current fix, but the userFloat is exchanged by a proper class member. If then the pat::Jet collection was not smeared before using it in the MET module, my approach would be to just not do anything besides what is done in the previous implementation. This just means that in this case only JES corrections exist and are reverted. This gives the responsibility to the user of the module to make sure that if smeared MET is desired, also smeared jets have to be put into the module (of course this is done already automatically by the corresponding MET python functions).

Regarding the reco::PFJet and reco::CaloJet classes, I'm not sure if we want to also add a member for these classes at the moment. The reason for that is that I'm not 100% convinced that the PFJetMETcorrInputProducerT.h actually handles the corrections correctly for these types. This is still something I'm trying to figure out.

So if this strategy would be good with you, I would change the PR accordingly (or should I create a new one?) and then I could come back to the reco jets later with another PR. This would "cure" MiniAOD, because there I think everything is working with pat::Jets anyway, and the rest will stay untouched.

Cheers
Michael

Comment (jpata) [reply PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc:32]: OK, I think I understand, and agree.

**Comment by michaelwassmer** [comment]
I converted the PR to draft for now, because I get some "undefined reference" errors from the linker associated with the following modules:
PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc
PhysicsTools/PatUtils/plugins/SmearedJetProducer.cc
However, I don't really understand why. My guess is that it has something to do with a buildfile. When I put the definition of the four functions https://github.com/cms-sw/cmssw/blob/3351a2b7b3797e77e948e2335bf2392fbcb6ffab/DataFormats/PatCandidates/src/Jet.cc#L620-L645 directly into the declaration in https://github.com/cms-sw/cmssw/blob/3351a2b7b3797e77e948e2335bf2392fbcb6ffab/DataFormats/PatCandidates/interface/Jet.h#L182-L192, the errors disappear. At this point I'm not really sure what to look at because I made sure that the description of the methods are similar in the .h and .cc file so maybe someone can give me a hint what the problem is.

**Comment by jpata** [comment]
dummy question, but did you try

```
scram b clean
scram b vclean
```

**Comment by jpata** [comment]
Also, `git cms-checkdeps -a` to make sure the dataformat update is propagated to dependent packages.

**Comment by michaelwassmer** [comment]
Thanks for the suggestions. Unfortunately I still get the same error. Just to make sure, I started fresh from a recent build and did the following:
```
scram project CMSSW_12_2_X_2021-11-29-1100
cd CMSSW_12_2_X_2021-11-29-1100/src
cmsenv
git-cms-rebase-topic michaelwassmer:SmearMETFix_rebased
scram b distclean
scram b clean
scram b vclean
# a lot of cleans I know, but wanted to be sure
git cms-checkdeps -a -A
scram b -j 8
```
However, still the same error.

**Comment by michaelwassmer** [comment]
Would it be possible to start the next check to see whether it works if everything is built from scatch?

**Comment by jpata** [comment]
@cmsbuild please test

(sorry, I thought you still need to check something on your side)

**Comment by jpata** [comment]
@michaelwassmer  I think the following is relevant and requires a new commit

```
>> Checking EDM Class Version for src/DataFormats/PatCandidates/src/classes_def_objects.xml in libDataFormatsPatCandidates.so
error: class 'pat::Jet' has a different checksum for ClassVersion 17. Increment ClassVersion to 18 and assign it to checksum 1291486363
Suggestion: You can run 'scram build updateclassversion' to generate src/DataFormats/PatCandidates/src/classes_def_objects.xml.generated with updated ClassVersion
```

**Comment by michaelwassmer** [comment]
> @michaelwassmer I think the following is relevant and requires a new commit
> 
> ```
> >> Checking EDM Class Version for src/DataFormats/PatCandidates/src/classes_def_objects.xml in libDataFormatsPatCandidates.so
> error: class 'pat::Jet' has a different checksum for ClassVersion 17. Increment ClassVersion to 18 and assign it to checksum 1291486363
> Suggestion: You can run 'scram build updateclassversion' to generate src/DataFormats/PatCandidates/src/classes_def_objects.xml.generated with updated ClassVersion
> ```

I added a new version to the xml (hopefully correctly, haven't done this before). With this I was actually able to built it locally without the "undefined reference" errors.

**Comment by jpata** [comment]
@cmsbuild please test

File: DataFormats/PatCandidates/interface/Jet.h [hunk 2]
```diff
@@ -650,6 +664,10 @@ namespace pat {
     // ---- id functions ----
     reco::JetID jetID_;
 
+    // ---- JER correction factor ----
+    float jerFactor;
```
Comment (jpata) [inline DataFormats/PatCandidates/interface/Jet.h:668]:
```suggestion
    float jerFactor_;
```

File: DataFormats/PatCandidates/interface/Jet.h [hunk 3]
```diff
@@ -650,6 +664,10 @@ namespace pat {
...
+    bool jerFactorValid;
```
Comment (jpata) [inline DataFormats/PatCandidates/interface/Jet.h:669]:
```suggestion
    bool jerFactorValid_;
```

File: DataFormats/PatCandidates/src/classes_def_objects.xml [hunk 4]
```diff
@@ -248,7 +248,8 @@
              newObj->setPflowIsolationVariables(pfIsoVar);]]>
   </ioread>
 
-  <class name="pat::Jet"  ClassVersion="17">
+  <class name="pat::Jet"  ClassVersion="18">
+   <version ClassVersion="18" checksum="1291486363"/>
```
Comment (jpata) [inline DataFormats/PatCandidates/src/classes_def_objects.xml:252]: I think you also need to introduce a read rule as follows below
```xml
  <ioread sourceClass = "pat::Jet" version="[1-17]" targetClass="pat::Jet" source="" target="jerFactorValid_">
    <![CDATA[jerFactorValid_ = false;]]>
  </ioread>
```
to make sure this is correctly initialized when reading old versions of this class

File: DataFormats/PatCandidates/interface/Jet.h [hunk 5]
```diff
@@ -177,6 +177,20 @@ namespace pat {
     void scaleEnergy(double fScale) override { scaleEnergy(fScale, "Unscaled"); }
     void scaleEnergy(double fScale, const std::string& level);
 
+    // Methods to work with a JER correction factor
+    //
+    // is the JER correction factor valid, i.e. has it been written using the saveJerFactor method
+    bool isJerFactorValid() const;
+
+    // load the currently saved JER correction factor and check whether the factor was propely set before loading
+    float loadJerFactor() const;
```
Comment (jpata) [inline DataFormats/PatCandidates/interface/Jet.h:186]:
```suggestion
    float jerFactor() const;
```
may be clearer, also follows better the rule 2.11: http://cms-sw.github.io/cms_coding_rules.html

> For a getter method, using the value name is preferred, e.g. momentum().

File: DataFormats/PatCandidates/interface/Jet.h [hunk 6]
```diff
@@ -177,6 +177,20 @@ namespace pat {
...
+
+    // save a JER correction factor
+    void saveJerFactor(float jerFactor_);
```
Comment (jpata) [inline DataFormats/PatCandidates/interface/Jet.h:189]:
```suggestion
    void setJerFactor(float jerFactor_);
```
as above for coding rules

File: DataFormats/PatCandidates/interface/Jet.h [hunk 7]
```diff
@@ -177,6 +177,20 @@ namespace pat {
...
+
+    // reset a saved JER correction factor
+    void resetJerFactor();
```
Comment (jpata) [inline DataFormats/PatCandidates/interface/Jet.h:192]: I'm not quite sure where this is needed later?

**Review by jpata — COMMENTED** [review]
Thank you for the update. Some comments inline.

I see reco differences in patMETs_slimmedMETs corrections and uncertainties: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-12-02-2300+6a446a/47223/validateJR/all_OldVSNew_TTbarFS13in2018MINIAODwf2018p1/
Screenshot from 2021-12-07 14-32-37

In the remini workflows 136.X, I only see differences in the uncertainties, not corrections: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-12-02-2300+6a446a/47223/validateJR/all_mini_OldVSNew_RunSinglePh2016Bwf136p731/
Screenshot from 2021-12-07 14-33-03

For my understanding, is this expected / by design?

The dd4hep / .911 workflow differences are unrelated, I think.

File: DataFormats/PatCandidates/interface/Jet.h (hunk 2 above)
Comment (michaelwassmer) [reply DataFormats/PatCandidates/interface/Jet.h]: done

File: DataFormats/PatCandidates/interface/Jet.h (hunk 3 above)
Comment (michaelwassmer) [reply DataFormats/PatCandidates/interface/Jet.h]: done

File: DataFormats/PatCandidates/interface/Jet.h (hunk 5 above)
Comment (michaelwassmer) [reply DataFormats/PatCandidates/interface/Jet.h]: done

File: DataFormats/PatCandidates/interface/Jet.h (hunk 6 above)
Comment (michaelwassmer) [reply DataFormats/PatCandidates/interface/Jet.h]: done

File: DataFormats/PatCandidates/interface/Jet.h (hunk 7 above)
Comment (michaelwassmer) [reply DataFormats/PatCandidates/interface/Jet.h:192]: At the moment this is not needed. I just added it for convenience. In the new commit, it is still included, however I can also remove it if desired.

File: DataFormats/PatCandidates/src/classes_def_objects.xml (hunk 4 above)
Comment (michaelwassmer) [reply DataFormats/PatCandidates/src/classes_def_objects.xml]: done

**Comment by michaelwassmer** [comment]
Thanks for the comments above. Regarding the observed differences, I will have another look and comment again, hopefully today.

**Comment by jpata** [comment]
@cmsbuild please test

File: DataFormats/PatCandidates/interface/Jet.h (hunk 7 above)
Comment (jpata) [reply DataFormats/PatCandidates/interface/Jet.h:192]: I would just remove it to avoid confusion.

**Comment by michaelwassmer** [comment]
> Thank you for the update. Some comments inline.
> 
> I see reco differences in patMETs_slimmedMETs corrections and uncertainties: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-12-02-2300+6a446a/47223/validateJR/all_OldVSNew_TTbarFS13in2018MINIAODwf2018p1/ Screenshot from 2021-12-07 14-32-37
> 
I think these differences are expected and originate exactly from the changes. As you can see, the corrections 6-8 and the uncertainties 16-17 are affected. These are exactly the corrections/uncertainties changed by this PR because they are associated with smeared MET, see https://github.com/michaelwassmer/cmssw/blob/SmearMETFix_rebased/DataFormats/PatCandidates/interface/MET.h#L198-L200 and https://github.com/michaelwassmer/cmssw/blob/SmearMETFix_rebased/DataFormats/PatCandidates/interface/MET.h#L169-L170 .

File: DataFormats/PatCandidates/interface/Jet.h (hunk 7 above)
Comment (michaelwassmer) [reply DataFormats/PatCandidates/interface/Jet.h:192]: Hi again,
I now implemented it in a way to better keep track whether the jerFactor is already set or not. With this the user should be aware that he overwrites an already existing factor and can then decide whether this is correct or not. If so, the resetJerFactor method can be used to be able to overwrite. Hope this is better.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
assign @cms-sw/xpog-l2 

just in case, as it touches MINI in a nontrivial way

**Comment by jpata** [comment]
We discussed today at the reco meeting that it needs to be understood if there is a plan / need to support smeared MET in legacy miniaod, which would imply the need for a backport to 10_6_X.

@cms-sw/jetmet-pog-l2 @camclean

**Comment by jpata** [comment]
any reason to keep it draft?

**Comment by michaelwassmer** [comment]
> any reason to keep it draft?

No, I don't think so. Just forgot to convert it back.

**Comment by jpata** [comment]
Thanks. From the reco side no additional questions at the moment.

For JME: We should wait for the confirmation if this needs to be backported. Since it implies a change from the default behaviour, it needs to be confirmed that these quantities were indeed not used in the old release, otherwise, a modifier will need to be used to protect the change.

File: DataFormats/PatCandidates/src/classes_def_objects.xml [hunk 8]
```diff
@@ -261,6 +262,9 @@
    <version ClassVersion="11" checksum="4153489469"/>
    <version ClassVersion="10" checksum="3393361159"/>
   </class>
+  <ioread sourceClass = "pat::Jet" version="[1-17]" targetClass="pat::Jet" source="" target="jerFactorValid_">
```
Comment (jpata) [inline DataFormats/PatCandidates/src/classes_def_objects.xml:265]: sorry, I had another chat with Slava, and I understand the read rule is not required (and thus should not be used), as the default value `jerFactorValid_(false)` is set in the constructor. Would you mind removing the read rule?

Sorry, this root IO business is not very straightforward.

Comment (michaelwassmer) [reply DataFormats/PatCandidates/src/classes_def_objects.xml:265]: Sure, no problem. Just a moment.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by errai-** [comment]
@michaelwassmer @jpata I was directed here from the newly introduced [critical bugfix](https://github.com/cms-sw/cmssw/pull/36559), as we have both edited the .cc file of pat::Jet. This fix seems to be a very welcome one, but I still needed to go through the whole code patch to check that there are no underlying conflicts.

What I'm slightly worried about here is that quite a lot of new code have been introduced for the seemingly simple problem of undoing and redoing the jer smearing. If I understood correctly, the main point is to be able to undo the smearing scale after jet.correctedP4("Uncorrected") is called in this [code](https://github.com/michaelwassmer/cmssw/blob/a1851de451e116e63d7e22d04872e3625ece7f2b/PhysicsTools/PatUtils/plugins/PATPFJetMETcorrInputProducer.cc#L27-L34).

Has it been recognized that this issue could have been solved with lots of less new code by simply calling jet.correctedP4("Unscaled") instead of jet.correctedP4("Uncorrected") and leaving the new parts with jet.jerFactor() out? From 10_6_X onwards, the [pat::Jet::scaleEnergy function](https://github.com/michaelwassmer/cmssw/blob/a1851de451e116e63d7e22d04872e3625ece7f2b/DataFormats/PatCandidates/src/Jet.cc#L267-L274) appends any additional jet energy scales by the title "Unscaled" in to the beginning of the JEC level vector. This function is most of all utilized by the [jet smearing routines](https://github.com/michaelwassmer/cmssw/blob/a1851de451e116e63d7e22d04872e3625ece7f2b/PhysicsTools/PatUtils/interface/SmearedJetProducerT.h#L106).

In short: if smearing is applied to a pat::Jet using the scaleEnergy function, one can undo the JECs and JER smearing simultaneously by referring to the "JEC level" "Unscaled" instead of "Uncorrected". In addition, if any additional energy scales have been applied, referring to "Unscaled" in correctedP4() automatically refers to the point where all energy scales are undone. In this sense, using "Unscaled" and the already existing structures is a more universal solution than the one proposed here. The only "additional" check that would be needed, is checking if the "lowest" available correction level is "Unscaled" or "Uncorrected".

Tagging also @ahinzmann , who was involved in introducing the "Unscaled" correction level in the first place.

**Comment by michaelwassmer** [comment]
Hi @errai- 
I wasn't aware of this "Unscaled" correction level. It seems that it would become a much simpler fix and I don't mind trying it out. I found the code where the "Unscaled" is introduced, however I couldn't quickly find where the "Uncorrected" level is created. Would you mind pointing me to it? Nevertheless, do I understand correctly that with the "Unscaled" level it is possible to revert all applied energy scalings and with the "Uncorrected" only the registered JEC levels?

**Comment by errai-** [comment]
> Hi @errai- I wasn't aware of this "Unscaled" correction level. It seems that it would become a much simpler fix and I don't mind trying it out. I found the code where the "Unscaled" is introduced, however I couldn't quickly find where the "Uncorrected" level is created. Would you mind pointing me to it? Nevertheless, do I understand correctly that with the "Unscaled" level it is possible to revert all applied energy scalings and with the "Uncorrected" only the registered JEC levels?

Hi @michaelwassmer and thanks for the prompt reply! The level "Uncorrected" is introduced when the pat::Jet is created in association to the jet energy corrections. Actually I'm not exactly sure where the code for this is located (maybe @ahinzmann can help?). I think it's best to give a brief example: take a pat::Jet where the JEC level is set to L3Absolute. The jet corrections (e.g. for correctedP4()) are all called through the function jecFactor, which looks for the corrections within an ordered vector. Simplifying a bit, as the jet is created, the ordered return values of jecFactor look like this:

Uncorrected=0.959391, L1FastJet=0.944361, L2Relative=1, L3Absolute=1

If one would like to swap e.g. to the level L1FastJet, one would need to multiply the current P4 by 0.944361. Here, the level Uncorrected (still) refers to the completely raw jet. The jer smearing is "postulated" onto a jet which already has the jet corrections. After the smearing, the (vector) of jecFactor return values will look like this:

Unscaled=0.936635 Uncorrected=0.959391 L1FastJet=0.944361 L2Relative=1 L3Absolute=1

The factor of the smearing factor is added into the beginning of this vector! This means that the meaning of "Uncorrected" is changed: it now refers to the jet with no jet corrections, but smearing is still applied. Calling "Unscaled" gives the completely uncorrected and unsmeared jet. Just as a reminder, it is also possible that scaleEnergy is called in other purposes than jet smearing. If one would call the function with a unitary dummy scale before smearing, one would end up with

Unscaled=0.936635 Unscaled=0.959391 Uncorrected=0.959391 L1FastJet=0.944361 L2Relative=1 L3Absolute=1

That is, there is a dummy "Unscaled" on the left-hand-side of "Uncorrected". However, calling correctedP4("Unscaled") will return the leftmost "Unscaled" case. Continuing with these guidelines, one could replace [this code reapplying the jer scale on a raw jet](https://github.com/michaelwassmer/cmssw/blob/a1851de451e116e63d7e22d04872e3625ece7f2b/JetMETCorrections/Type1MET/interface/PFJetMETcorrInputProducerT.h#L210-L212) by adding something like the following into the [jetcorrextractor file](https://github.com/michaelwassmer/cmssw/blob/a1851de451e116e63d7e22d04872e3625ece7f2b/JetMETCorrections/Type1MET/interface/JetCorrExtractorT.h):

```
if constexpr (std::is_same<T, pat::Jet>::value) {
  rawJetP4 *= rawJet.jecFactor("Uncorrected)/rawJet.("Unscaled");
}
```

**Comment by camclean** [comment]
@jpata @michaelwassmer I have confirmed with @kirschen that we would like this fix to be available for UL analyses and therefore should also have a backport.

**Comment by jpata** [comment]
@camclean thanks for the confirmation.

@michaelwassmer @errai- thanks for the discussion on the implementation. I'm not an expert on JME corrections code as such, so I leave it to JME to discuss thoroughly and propose the best implementation. Clearly, if something related ("Unscaled") already exists and we can avoid introducing more layers of complexity, it should be preferred. I could suggest to discuss this next year with all parties involved (at e.g. the JME meeting) before considering this PR for merging.

**Comment by errai-** [comment]
> @camclean thanks for the confirmation.
> 
> @michaelwassmer @errai- thanks for the discussion on the implementation. I'm not an expert on JME corrections code as such, so I leave it to JME to discuss thoroughly and propose the best implementation. Clearly, if something related ("Unscaled") already exists and we can avoid introducing more layers of complexity, it should be preferred. I could suggest to discuss this next year with all parties involved (at e.g. the JME meeting) before considering this PR for merging.

Thanks for the comment @jpata . I gave @kirschen a quick ping about the topic, in case he hasn't started his holidays yet.

**Comment by jpata** [comment]
kind ping on this. will there be a discussion in JME about the implementation?

**Comment by ahinzmann** [comment]
Just my 2 cents. I have no strong opinion whether to implement a dedicated interface for JER in pat::Jet, or keep the current (which uses "unscaled"). One way to address the bug may to replace correctedP4("uncorrected") by correctedP4(0) where necessary. This appears in several places in CMSSW and it's obvious, also how to keep MiniAOD and NanoAOD in sync:
https://github.com/cms-sw/cmssw/search?q=correctedP4

**Comment by jpata** [comment]
kind ping @michaelwassmer, has there been any discussion on the path forward in JME?

**Comment by michaelwassmer** [comment]
Hi,
sorry for being so silent the last days. There hasn't been much of a discussion, but I think @errai- has suggested the most elegant solution. Whether to use correctedP4("unscaled") or correctedP4(0) is a matter of personal taste I guess. But these suggestions are both better than what I did earlier, since they use already existing methods. I would suggest that we go with "unscaled" for now. I think I can implement this in the next few days. Should I create an new PR for this or just keep working on this one?

**Comment by errai-** [comment]
> Hi, sorry for being so silent the last days. There hasn't been much of a discussion, but I think @errai- has suggested the most elegant solution. Whether to use correctedP4("unscaled") or correctedP4(0) is a matter of personal taste I guess. But these suggestions are both better than what I did earlier, since they use already existing methods. I would suggest that we go with "unscaled" for now. I think I can implement this in the next few days. Should I create an new PR for this or just keep working on this one?

Just as a quick comment: there might be a slight preference towards @ahinzmann 's correctedP4(0) for two reasons:
- It's completely future-proof, if additional layears are to be added
- If unsmeared jets are used, the lack of an 'Unscaled' layer does not produce issues
Admittedly, correctedP4(0) is not very readable/clean code, so that's the downside one would need to accept.

**Comment by michaelwassmer** [comment]
Thanks @errai- 
All right, then I would go with correctedP4(0) to obtain the completely uncorrected jet so I don't have to manually use the smearing factor to reverse the smearing correction. I completely agree with this.
However, I have one further question:
The JetMETCorrections/Type1MET/interface/PFJetMETcorrInputProducerT.h module propagates the Type1 MET corrections by getting the completely uncorrected jets (this is clear, see above) and then recalculates from them the
- jets at L1 correction
- and the jets at the desired correction level you provide in the corresponding python config (usually L1L2L3).

Then from the residuals of the L1 collection and the L1L2L3 collection it calculates the Type1 MET corrections.
So far so good. For the smeared MET, the only difference is that one needs to consider the smearing corrections on top of the L1L2L3 jet collection. The only problem I currently see is that

`    rawJetP4 *= rawJet.jecFactor("Uncorrected)/rawJet.jecFactor("Unscaled");`

implicitly assumes that the only "difference" between jecFactor("Uncorrected") and jetFactor("Unscaled") (or jecFactor(0)) is the smearing correction. I guess this could be problematic if for any reason the scaleEnergy method was invoked in addition to the one time during the smearing procedure because then one would not be able to get the smearing factor from this ratio. Am I missing something?
If not, I think it could make sense to still have the JER correction interface to keep track of the JER factor or the JER would need to be registered as an additional JEC in the already exiting jecFactors interface. Does this make sense?

**Comment by errai-** [comment]
@michaelwassmer all scaleEnergy calls go into the same 'box' with jet smearing, accumulating the total jet energy scale. That is, if one adds extra scales to jets e.g. for the purpose of systematic variations, one needs to consider them in MET calculus exactly in the same manner as the jet smearing scales. Thus for MET, I see no reason of separating smearing scales from other scaleEnergy call types. One could think of all the non-JEC scales as one single generic extra scale that needs to be considered. Does this answer your question, or is there some subtle detail I missed?

Finally, to underline the features of the code: the level "Uncorrected" always exists, but "Unscaled" only exists if scaleEnergy has been called e.g. in the purpose of jet smearing. Hence, jecFactor(0) is the safest and most universal call one can make, as it corresponds to "Uncorrected" in the cases where "Unscaled" is not available for some reason.

**Comment by michaelwassmer** [comment]
I created a new PR (https://github.com/cms-sw/cmssw/pull/36797) in which I incorporated the suggested changes. I thought it would probably be cleaner to have a new PR since this one is already quite long.

**Comment by jpata** [comment]
-reconstruction

**Comment by jpata** [comment]
@cmsbuild please close
